### ISSUE TITLE
[RL algorithm bug?] SAC actor loss

### ISSUE THREAD

**Issue by Kenneth-Schroeder** [aai-institute/tianshou#489]
Hi all,
I was just looking into the [SAC code](https://github.com/thu-ml/tianshou/blob/3592f45446e6cc98423df2f1c28d8ca0ef2be821/tianshou/policy/modelfree/sac.py#L165) and the entropy term seems a bit off to me.
As far as I can tell, in your implementation Q-values are don't contain an entropy term and state-values are not used.
The actor loss is computed as follows:

```
actor_loss = (
    self._alpha * obs_result.log_prob.flatten() -
    torch.min(current_q1a, current_q2a)
).mean()
```

... where alpha is the entropy coefficient and q1a/q2a are the critics q-value estimations.
It seems to me like the raw log probabilities are used as optimization objective here,
which should mean that low probabilities (= 0%) are very much encouraged by the algorithm,
except for one event with probability 100% and a large Q-value.
As I understand [SAC](https://arxiv.org/pdf/1801.01290.pdf), this is exactly, what the algorithm tries to avoid.
The goal should be to achieve as much randomness as possible, i.e. high entropy.

My understanding tells me that the actor loss should instead be computed using x log x, something around these lines:

```
actor_loss = (
    self._alpha * obs_result.prob.flatten() * obs_result.log_prob.flatten() -
    torch.min(current_q1a, current_q2a)
).mean()
```

Feedback is appreciated and I am happy to be corrected :)

**Comment by danagi** [comment]
Hi.
Our implementation is based on the new verison (arXiv:1812.05905.) rather than the old one (arXiv: 1801.01290). In this version, state-values are no longer needed.
The (differential) entropy of a continuous random variable with density f(x) is defined as `∫ f(x) log f(x)`. Here, SAC estimates the entropy term by `log pi(a|s)` where `a` is a action sampled by reparameterization trick.

**Comment by danagi** [comment]
Hi.
Our implementation is based on the new verison (arXiv:1812.05905.) rather than the old one (arXiv: 1801.01290). In this version, state-values are no longer needed.
The (differential) entropy of a continuous random variable with density f(x) is defined as `∫ f(x) log f(x)`. Here, SAC estimates the entropy term by `log pi(a|s)` where `a` is a action sampled by reparameterization trick.

**Comment by Kenneth-Schroeder** [comment]
Thank you! I had some catching up to do. Now I understand much better :)
